### PR TITLE
Make SpinReversalTransformComposite nonblocking

### PR DESCRIPTION
**Pull request by arcondello**
Also did a refactor to use more modern Ocean syntax and fixed a bunch of pep8 issues.

Closes https://github.com/dwavesystems/dwave-preprocessing/issues/111

### CODE REVIEW

**Review by jackraymond — APPROVED** [review]
Seed implementation and testing is looks good, other code looks fine but not too familiar with blocking.

File: dwave/preprocessing/composites/spin_reversal_transform.py [hunk 1]
```diff
@@ -91,38 +133,52 @@ def sample(self, bqm, *, num_spin_reversal_transforms=1, **kwargs):
...
+
+        # Get the SRT matrix
+        SRT = self.rng.random((num_spin_reversal_transforms, bqm.num_variables)) > .5
+
+        # Submit the problems
+        samplesets: typing.List[dimod.SampleSet] = []
+        flipped = np.zeros(bqm.num_variables, dtype=bool)  # what variables are currently flipped
+        for i in range(num_spin_reversal_transforms):
+            # determine what needs to be flipped
+            transform = flipped != SRT[i, :]
```
Comment (randomir) [inline dwave/preprocessing/composites/spin_reversal_transform.py:160]: Clever trick to minimize flips.

We could further optimize it by sorting the `SRT` by Hamming distance, or in "gray code order".

**Review by randomir — APPROVED** [review]
LGTM

File: dwave/preprocessing/composites/spin_reversal_transform.py [hunk 2]
```diff
@@ -52,31 +53,72 @@ class SpinReversalTransformComposite(Sampler, Composite):
...
-        parameters.update(child.parameters)
-
-        self.properties = {'child_properties': child.properties}
-
-    def sample(self, bqm, *, num_spin_reversal_transforms=1, **kwargs):
+    _children: typing.List[dimod.core.Sampler]
+    _parameters: typing.Dict[str, typing.Sequence[str]]
+    _properties: typing.Dict[str, typing.Any]
+
+    def __init__(self, child: dimod.core.Sampler, *, seed=None):
```
[the remark this answers is not in the document]
Comment (arcondello) [reply dwave/preprocessing/composites/spin_reversal_transform.py:60]: The typing for seed is actually a mess, and ends up making the function signature in the documentation completely unreadable. So figured for brevity I would leave it alone.

File: dwave/preprocessing/composites/spin_reversal_transform.py [hunk 3]
```diff
@@ -52,31 +53,72 @@ class SpinReversalTransformComposite(Sampler, Composite):
...
+        return parameters
+
+    @property
+    def properties(self) -> typing.Dict[str, typing.Any]:
+        try:
+            return self._properties
+        except AttributeError:
+            pass
+
+        self._properties = dict(child_properties=self._child.properties)
```
[the remark this answers is not in the document]
Comment (arcondello) [reply dwave/preprocessing/composites/spin_reversal_transform.py:92]: Hmm, I think there is an argument either way. We didn't in the old version of the code. My inclination is just to blindly inherit. If something was to modify the properties of the child sampler I guess we'd want that information propagated out.

File: dwave/preprocessing/composites/spin_reversal_transform.py [hunk 4]
```diff
@@ -52,31 +53,72 @@ class SpinReversalTransformComposite(Sampler, Composite):
...
-
-        self.properties = {'child_properties': child.properties}
-
-    def sample(self, bqm, *, num_spin_reversal_transforms=1, **kwargs):
+    _children: typing.List[dimod.core.Sampler]
+    _parameters: typing.Dict[str, typing.Sequence[str]]
+    _properties: typing.Dict[str, typing.Any]
+
+    def __init__(self, child: dimod.core.Sampler, *, seed=None):
+        self._child = child
```
[the remark this answers is not in the document]
Comment (arcondello) [reply dwave/preprocessing/composites/spin_reversal_transform.py:61]: This is one of those places where we very often do duck typing. E.g. a lot of the tests rely on that. I'm inclined to leave it be.

File: dwave/preprocessing/composites/spin_reversal_transform.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arcondello) [reply dwave/preprocessing/composites/spin_reversal_transform.py:60]: Ah, excellent catch

**Pull request merged**